neuro_evolution_ctrnn/tools/algorithms.py

```diff
@@ -69,6 +70,31 @@ def record_individuals(toolbox, population):
                                                        k=toolbox.conf.novelty.recorded_behaviors_per_generation)
 
 
+def validation_runs(population, toolbox):
+    intial_seed = toolbox.conf.validation_initial_seed
+    number_validation_runs = toolbox.conf.validation_number_runs
+
+    if number_validation_runs < 1:
+        return 0, 0
+    seeds_validation_runs: np.ndarray = np.array(range(intial_seed, intial_seed + number_validation_runs))
+    best_individuals = tools.selBest(population, 1, fit_attr="fitness")  # only validate the single best individual
```

Comment:
Hmm sollte man die Anzahl die hier gewählt wird als Parameter machen? Z.B. wenn `random_seed=-1` gewählt wird und `toolbox.conf.fix_seed_for_generation` auf False dann bekommt jedes Individuum potentiell einen anderen Seed bei der Evaluation oder? Dann wäre das Ergebnis (also die Fitness) des besten Individuums in dieser Zeile auch stark an den Seed der Environment gebunden. Könnte ja sein dass das beste Individuum gemessen an der Fitness nur gut war weil die Environment einfach war und nicht weil das Individuum besser war. Dann wäre es gut z.B. noch die anderen der Top5 Individuen zu nehmen und die zu testen oder was meinst du?

Natürlich muss man irgendwo eine Grenze ziehen, man kann ja nicht alle nochmal neu evaluieren

Reply:
Ich hab gerade https://github.com/neuroevolution-ai/NeuroEvolution-CTRNN_new/pull/70#issuecomment-775941476 gesehen. Vielleicht liegt das hier dran

Reply:
>  False dann bekommt jedes Individuum potentiell einen anderen Seed bei der Evaluation ode

ne. Der Episoderunner kennt nur den Seed, der beim aufrufe von `eval_fitness` mitgegeben wird. Also im Falle von validation runs, ist das immer `seeds_validation_runs`. Die Sache mit dem `fix_seed_for_generation` und dem `random_seed` beeinflusst welchen seed der Optimizer dem EpisodeRunner mitgibt. Und daher sollten beide parameter keinen einfluss auf die validationruns haben.

Reply:
Also mir ging es nicht um den Einfluss auf die Validation Runs sondern die Art wie Individuen ausgewählt werden für die Validation Runs.

Bei den Seeds bezieh ich mich auf die Evaluationen (oder hatten wir das jetzt Trainings Runs genannt? jedenfalls die die wir mit jedem Individuum der Population durchführen). Wenn wir da für jedes Invdividuum einen anderen Seed haben sollten (das passiert glaube ich wenn `random_seed=-1` und `toolbox.conf.fix_seed_for_generation = False`), dann könnte es ja sein dass manche Individuen besser sind als andere aber da sie einen schwereren Seed bekommen haben eine weniger gute Fitness bekommen. Dann würden sie hier untergehen weil nur das Beste Individuum genommen wird (auch wenn dieses vielleicht einen leichten Seed hatte und dadurch eine gute Fitness erzielt hat).

Wenn man einen Parameter `number_best_individuals: int` einführt könnte man wengistens z.B. die 5 Besten nehmen für die Validation Runs.